### PR TITLE
Adding `all_different` and `count_if` to the logical expression system

### PR DESCRIPTION
**Pull request by emma58**
## Summary/Motivation:

This adds two nodes to the logical expression system: `all_different` and `count_if`. It also adds handling for these nodes in the CP Optimizer writer in `contrib.cp`.

Note: The tests are passing, but they hit #3057 out the wazoo

## Changes proposed in this PR:
- Adds `all_different`
- Adds `count_if`
- Adds handling for these in the docplex CP model writer.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/core/expr/logical_expr.py [hunk 1]
```diff
@@ -512,4 +531,56 @@ def _apply_operation(self, result):
...
+
+
+class CountIfExpression(NumericExpression):
+    """
+    Logical expression that returns the number of True child statements.
+    All arguments are expected to be Boolean-valued.
+    """
+
+    __slots__ = ()
+    PRECEDENCE = 10  # TODO: maybe?
```
Comment (mrmundt) [inline pyomo/core/expr/logical_expr.py:564]: Any update on the `maybe?` here?

**Review by mrmundt — APPROVED** [review]
I have one question but from a quality perspective, looks good. Lots of tests. I love tests.

File: pyomo/core/expr/logical_expr.py [hunk 2]
```diff
@@ -240,6 +239,26 @@ def atleast(n, *args):
...
 
 
+def all_different(*args):
+    """Creates a new AllDifferentExpression
+
+    Requires all of the arguments to take on a different value
+
+    Usage: all_different(m.X1, m.X2, ...)
+    """
+    return AllDifferentExpression(list(_flattened(args)))
```
Comment (jsiirola) [inline pyomo/core/expr/logical_expr.py:249]: What verifies that all arguments are NumericExpressions?

File: pyomo/core/expr/logical_expr.py [hunk 3]
```diff
@@ -240,6 +239,26 @@ def atleast(n, *args):
...
+
+
+def count_if(*args):
+    """Creates a new CountIfExpression
+
+    Counts the number of True-valued arguments
+
+    Usage: count_if(m.Y1, m.Y2, ...)
+    """
+    return CountIfExpression(list(_flattened(args)))
```
Comment (jsiirola) [inline pyomo/core/expr/logical_expr.py:259]: What verifies that the arguments are all LogicalExpressions?

File: pyomo/core/expr/logical_expr.py [hunk 4]
```diff
@@ -512,4 +531,56 @@ def _apply_operation(self, result):
...
+    """
+
+    __slots__ = ()
+    PRECEDENCE = 10  # TODO: maybe?
+
+    def __init__(self, args):
+        # require a list, a la SumExpression
+        if args.__class__ is not list:
+            args = list(args)
+        self._args_ = args
```
Comment (jsiirola) [inline pyomo/core/expr/logical_expr.py:570]: Why is this needed for CountIf and not for AllDiff?

**Review by jsiirola — CHANGES_REQUESTED** [review]
Over all, this is in good shape, but there are a couple minor things that should probably be addressed prior to merging.

File: pyomo/core/expr/logical_expr.py [hunk 5]
```diff
@@ -512,4 +531,56 @@ def _apply_operation(self, result):
...
+        return len(self._args_)
+
+    def getname(self, *arg, **kwd):
+        return 'count_if'
+
+    def _to_string(self, values, verbose, smap):
+        return "count_if(%s)" % (", ".join(values))
+
+    def _apply_operation(self, result):
+        return sum(value(r) for r in result)
```
[the remark this answers is not in the document]
Comment (emma58) [reply pyomo/core/expr/logical_expr.py:583]: I guess not?

File: pyomo/core/expr/logical_expr.py (hunk 4 above)
Comment (emma58) [reply pyomo/core/expr/logical_expr.py:570]: Whoops, it's not. I don't need to implement my own `__init__` here--just removed it.

File: pyomo/core/expr/logical_expr.py (hunk 2 above)
Comment (emma58) [reply pyomo/core/expr/logical_expr.py]: I just added a numeric and Boolean version of `_flattened` that will check for these and the other logical expressions.

File: pyomo/core/expr/logical_expr.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (emma58) [reply pyomo/core/expr/logical_expr.py]: Whoops, I was wrong. Yes, it is necessary, because `r` can be a relational expression (or anything Boolean-valued), so we need to evaluate it.

[the remark this answers is not in the document]
Comment (jsiirola) [reply pyomo/core/expr/logical_expr.py:631]: Shouldn't those expressions have already been evaluated at lower levels in the expression tree?  This should be called on the way up, so `result` should only have the results from evaluating the subexpressions.

File: pyomo/core/expr/logical_expr.py [hunk 6]
```diff
@@ -10,9 +10,8 @@
 #  This software is distributed under the 3-clause BSD License.
 #  ___________________________________________________________________________
 
-
 import types
-from itertools import islice
+from itertools import combinations, islice
```
Comment (jsiirola) [inline pyomo/core/expr/logical_expr.py:14]: Combinations can be removed (the code that used it was deleted)

File: pyomo/core/expr/logical_expr.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jsiirola) [reply pyomo/core/expr/logical_expr.py:631]: OK - this is an issue with how `GetItemExpression` (and their ilk) interact with the evaluation walker.  I am working on fixing that - which will be resolved in a separate PR.
